### PR TITLE
CATROID-1391 Cloudflare client certificate implemented

### PR DESCRIPTION
**Pull request by dorianpercic**
https://jira.catrob.at/browse/CATROID-1391
Cloudflare certificate inserted into each request using the CatrobatWebClient and RetrofitWebServer. Also unit test implemented, to test implementation with a test route. 

- Edit 20.10.2022: org.catrobat.catroid.test.web.CloudflareCertificateTest.testRouteUnsuccessful fails on Jenkins, because we always get back 200 from the server, because the emulator is maybe in the same network like the server. Maybe test will be removed.

- Edit 10.07.2023:
In my opinion and in opinion of others, the ticket shall be closed, as this does not provide a secure solution to the problem. Solution should be on the backend-side.

### Your checklist for this pull request
Please review the [contributing guidelines](https://github.com/Catrobat/Catroid/blob/develop/README.md) and [wiki pages](https://github.com/Catrobat/Catroid/wiki/) of this repository.

- [x] Include the name of the Jira ticket in the PR’s title
- [x] Include a summary of the changes plus the relevant context
- [x] Choose the proper base branch (*develop*)
- [x] Confirm that the changes follow the project’s coding guidelines
- [x] Verify that the changes generate no compiler or linter warnings
- [x] Perform a self-review of the changes
- [x] Verify to commit no other files than the intentionally changed ones
- [x] Include reasonable and readable tests verifying the added or changed behavior
- [x] Confirm that new and existing unit tests pass locally
- [x] Check that the commits’ message style matches the [project’s guideline](https://github.com/Catrobat/Catroid/wiki/Commit-Message-Guidelines)
- [x] Stick to the project’s gitflow workflow
- [x] Verify that your changes do not have any conflicts with the base branch
- [x] After the PR, verify that all CI checks have passed
- [x] Post a message in the *catroid-stage* or *catroid-ide* [Slack channel](https://catrobat.slack.com) and ask for a code reviewer

### CODE REVIEW

**Review by friendo97 — APPROVED** [review]
done in PP!

File: cloudflare.properties
Comment (moberer) [inline cloudflare.properties:1]: Never add a secret key to a git repo.

File: catroid/src/main/java/org/catrobat/catroid/web/CatrobatWebClient.kt [hunk 1]
```diff
@@ -23,16 +23,58 @@
...
+const val SECRET = BuildConfig.CLOUDFLARE_SECRET
+const val LINE_LENGTH = 64
+
+private fun setUpClient(): OkHttpClient {
+    val certificateFactory: CertificateFactory = CertificateFactory.getInstance("X.509")
+    val keySpec = getPrivateKey()
+
+    val certificate: Certificate? = getCertificate(certificateFactory)
+
+    val keyFactory: KeyFactory = KeyFactory.getInstance("RSA")
```
Comment (moberer) [inline catroid/src/main/java/org/catrobat/catroid/web/CatrobatWebClient.kt:64]: Why use RSA?

Prefer elliptic curves! - Faster and more future-proof.

File: catroid/src/main/java/org/catrobat/catroid/web/CatrobatWebClient.kt [hunk 2]
```diff
@@ -65,3 +105,67 @@ fun Map<String, String>.createFormEncodedRequest(url: String): Request {
         .post(formEncodingBuilder.build())
         .build()
 }
+
+private fun getPrivateKey(): PKCS8EncodedKeySpec {
+    val privateKeyInputStream: InputStream = BuildConfig.CLOUDFLARE_PRIVATE_KEY.byteInputStream()
+    val privateKeyByteArray =
+        ByteArray(privateKeyInputStream.available())
+    privateKeyInputStream.read(privateKeyByteArray)
+    val privateKeyContent = String(privateKeyByteArray, Charset.defaultCharset())
```
Comment (moberer) [inline catroid/src/main/java/org/catrobat/catroid/web/CatrobatWebClient.kt:114]: Do not manually mangle around with the key. There has to be a function to import a standard X509 certificate...


==> E.g. use the Java Keystore functionality?

**Review by moberer — CHANGES_REQUESTED** [review]

**Pull request closed without merging**